### PR TITLE
F catch handle error when tag name already exist

### PR DESCRIPTION
**Pull request by bakridemos**
Description: 

- adjust exception to be used while catched if any errors ( if the tag name exist already ).
- catch the exception and add a message bag to make the error more understandable to users: if there is an error while import ( tag name exist already ), the user will be able to see which tag name is not imported and why.
- add a new translation key

### How to review/test

-  csv Tags documents which have dupricated tags should not be able to be imported.

### CODE REVIEW

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php [hunk 1]
```diff
@@ -10,15 +10,41 @@
...
 namespace demosplan\DemosPlanCoreBundle\Exception;
 
+use demosplan\DemosPlanCoreBundle\Entity\Procedure\Procedure;
 use Exception;
+use demosplan\DemosPlanCoreBundle\Entity\Statement\TagTopic;
 
 class DuplicatedTagTitleException extends Exception
 {
+    public function __construct(string                       $message,
+                                protected readonly Procedure $procedure,
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php:20]: `Procedure` is not needed

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php [hunk 2]
```diff
@@ -10,15 +10,41 @@
...
-    public static function createFromTitleAndProcedureId(string $tagTitle, string $procedureId): self
+    public static function createFromTitleAndProcedureId(Procedure $procedure, TagTopic $topic, string $tagTitle): self
+    {
+        return new self("A tag with the title {$tagTitle} already exist in a procedure with the ID {$procedure->getId()}", $procedure, $topic, $tagTitle);
+    }
+
+    public function getProcedure(): Procedure
+    {
+        return $this->procedure;
+    }
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php:39]: `Procedure` is not needed

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php [hunk 3]
```diff
@@ -10,15 +10,41 @@
...
+        parent::__construct($message, 0, null);
+    }
+
     /**
      * @return static
      */
-    public static function createFromTitleAndProcedureId(string $tagTitle, string $procedureId): self
+    public static function createFromTitleAndProcedureId(Procedure $procedure, TagTopic $topic, string $tagTitle): self
+    {
+        return new self("A tag with the title {$tagTitle} already exist in a procedure with the ID {$procedure->getId()}", $procedure, $topic, $tagTitle);
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php:33]: `Procedure` is not needed

**Review by mussbach — CHANGES_REQUESTED** [review]

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php (hunk 1 above)
Comment (bakridemos) [reply demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php:20]: i ll fix it, i think i did another mistake here which i have to check, this way the message bag will be showed only while importing, the goal was to use it for the other case too ( while manual tag creating ).

Comment (bakridemos) [reply demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php]: fixed in https://github.com/demos-europe/demosplan-core/pull/2018/commits/6a030eb8aa1a3db49fa4be05a24639d2972c2fd9

File: demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php [hunk 4]
```diff
@@ -180,8 +181,8 @@ public function tagListEditAction(
                     $result = $statementHandler->createTagFromTopicId($requestPost['r_createtags'], $newTag, $procedure);
                     $this->getMessageBag()->add('confirm', 'confirm.tag.created');
                     $anchor = $result->getId();
-                } catch (Exception) {
-                    $this->getMessageBag()->add('warning', 'warning.tag.created');
+                } catch (DuplicatedTagTitleException $e) {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php:184]: What should happen when another exception like `BadRequestException` is thrown? It is not caught any more. Is this correct?

File: demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php [hunk 5]
```diff
@@ -317,8 +318,12 @@ public function tagListEditAction(
         // Check if we need to import tags
         if (\array_key_exists('r_import', $requestPost)) {
             if (\array_key_exists('r_importCsv', $requestPost) && '' != $requestPost['r_importCsv']) {
-                $statementHandler->importTags($procedure, $requestPost['r_importCsv']);
-                $this->getMessageBag()->add('confirm', 'explanation.import.topicsAndTags');
+                try {
+                    $statementHandler->importTags($procedure, $requestPost['r_importCsv']);
+                    $this->getMessageBag()->add('confirm', 'explanation.import.topicsAndTags');
+                } catch (DuplicatedTagTitleException $e) {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php:324]: Same as above

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php [hunk 6]
```diff
@@ -10,15 +10,33 @@
...
+use demosplan\DemosPlanCoreBundle\Entity\Statement\TagTopic;
 use Exception;
 
 class DuplicatedTagTitleException extends Exception
 {
+    public function __construct(string $message,
+                                protected readonly TagTopic $topic,
+                                protected readonly string $tagTitle,
+    ) {
+        parent::__construct($message, 0, null);
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php:22]: `0, null` is not needed as they are the default values

**Review by mussbach — CHANGES_REQUESTED** [review]

File: demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php (hunk 6 above)
Comment (bakridemos) [reply demosplan/DemosPlanCoreBundle/Exception/DuplicatedTagTitleException.php]: done thx https://github.com/demos-europe/demosplan-core/pull/2018/commits/98ea6dc40e8747273304f891b14de3eafe0dc9e0

File: demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php (hunk 5 above)
Comment (bakridemos) [reply demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php:331]: there are here other throwed exceptions which have to be handled, but BadRequestException i think is not one of them. all exceptions are anyway catched and handled now here https://github.com/demos-europe/demosplan-core/commit/5e0e05c33fbcc40c4b1965e8b4845441083c9eb3

File: demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php [hunk 7]
```diff
@@ -180,8 +182,14 @@ public function tagListEditAction(
...
-                } catch (Exception) {
-                    $this->getMessageBag()->add('warning', 'warning.tag.created');
+                } catch (DuplicatedTagTitleException $e) {
+                    $this->getMessageBag()->add('error', 'error.import.tag.name.taken', ['tagTitle' => $e->getTagTitle(), 'topicName' => $e->getTopic()->getTitle()]);
+                } catch (TagTopicNotFoundException $e) {
+                    $this->getMessageBag()->add('error', 'error.topic.notfound');
+                    $this->logger->error('error.topic.notfound', [$e]);
+                } catch (Exception $e) {
+                    $this->getMessageBag()->add('error', 'error.tag.add');
+                    $this->logger->error('error.tag.add', [$e]);
```
Comment (dresslerdemos) [inline demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php:192]: It doesn't make much difference, but why did you replace `warning.tag.created` with `error.tag.add`? The former one was a bit more specific.

**Review by dresslerdemos — CHANGES_REQUESTED** [review]

File: demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php (hunk 7 above)
Comment (bakridemos) [reply demosplan/DemosPlanCoreBundle/Controller/Statement/DemosPlanStatementTagController.php:192]: i did adjust it hier https://github.com/demos-europe/demosplan-core/pull/2018/commits/07c23fa1636376e951a988a945121b0443643e70, bit i think this an error message and not a warning ! in this case the tag will be not created while an error ( other than DuplicatedTagTitleException and TagTopicNotFoundException ) occurred

**Review by dresslerdemos — APPROVED** [review]
Code review only

**Pull request merged**